internal/provider/allocation.go

```diff
@@ -73,41 +114,147 @@ func (r *allocationResource) populateState(ctx context.Context, state *allocatio
 			// The resource was deleted. This is an edge case for create,
 			// but necessary for the read function.
 			state.Id = types.StringNull()
-			return d
+			return
 		}
-		d.AddError(
+		diags.AddError(
 			"Error Reading Doit Console Allocation",
 			"Could not read Doit Console Allocation ID "+state.Id.ValueString()+": "+err.Error(),
 		)
-		return d
+		return
 	}
 
 	state.Id = types.StringPointerValue(resp.Id)
 	state.Type = types.StringPointerValue(resp.Type)
-	state.Description = types.StringPointerValue(resp.Description)
+	if resp.Description != nil && *resp.Description != "" {
+		state.Description = types.StringPointerValue(resp.Description)
+	}
 	state.AnomalyDetection = types.BoolPointerValue(resp.AnomalyDetection)
 	state.CreateTime = types.Int64PointerValue(resp.CreateTime)
 	state.UpdateTime = types.Int64PointerValue(resp.UpdateTime)
 	state.Name = types.StringPointerValue(resp.Name)
+	state.UnallocatedCosts = types.StringPointerValue(resp.UnallocatedCosts)
 
 	if resp.AllocationType != nil {
 		state.AllocationType = types.StringValue(string(*resp.AllocationType))
 	}
 
 	if resp.Rule != nil {
-		var diags diag.Diagnostics
-		state.Rule.Formula = types.StringPointerValue(resp.Rule.Formula)
-		state.Rule.Components, diags = toAllocationRuleComponentsListValue[resource_allocation.ComponentsType](ctx, resp.Rule.Components)
-		d.Append(diags...)
-		if d.HasError() {
-			return d
+		m := map[string]attr.Value{
+			"formula": types.StringValue(resp.Rule.Formula),
+		}
+		if resp.Rule.Components != nil {
+			m["components"], d = toAllocationRuleComponentsListValue(ctx, resp.Rule.Components)
+			diags.Append(d...)
+			if diags.HasError() {
+				return
+			}
 		}
+		state.Rule, d = resource_allocation.NewRuleValue(resource_allocation.RuleValue{}.AttributeTypes(ctx), m)
+		diags.Append(d...)
+		if diags.HasError() {
+			return
+		}
+	} else {
+		state.Rule = resource_allocation.NewRuleValueNull()
 	}
 
-	return d
+	if resp.Rules != nil && len(*resp.Rules) > 0 {
+		// Map to store existing actions
+		existingActionsByID := make(map[string]string)
+		existingActionsByIndex := make([]string, 0)
+
+		if !state.Rules.IsNull() && !state.Rules.IsUnknown() {
+			var stateRules []resource_allocation.RulesValue
+			// We try to extract existing rules to preserve the "action" field which is not returned by the API.
+			// If this fails, we proceed without existing actions.
+			if d := state.Rules.ElementsAs(ctx, &stateRules, false); !d.HasError() {
+				for _, rule := range stateRules {
+					action := rule.Action.ValueString()
+					existingActionsByIndex = append(existingActionsByIndex, action)
+					if !rule.Id.IsNull() && !rule.Id.IsUnknown() {
```

Comment:
I haven't read the tests yet, but if not tested can you think of a reproducible use case to ensure we'll not bump into issues (permadiffs, etc)?